T4MVCExtensions/T4Extensions.cs

```diff
@@ -334,9 +334,11 @@ public static string JavaScriptReplaceableUrl(this UrlHelper urlHelper, ActionRe
             string action = token.ToString();
 
             var specificActionUrl = (from r in RouteTable.Routes.OfType<Route>()
-                                     where r.DataTokens != null && r.Defaults != null
-                                     && r.DataTokens.Any(dt => string.Compare(dt.Key, "area", StringComparison.InvariantCultureIgnoreCase) == 0
-                                         && string.Compare(dt.Value.ToString(), area, StringComparison.InvariantCultureIgnoreCase) == 0)
+                                     where r.Defaults != null
+                                     && ((string.IsNullOrEmpty(area) 
+                                        && (r.DataTokens == null || !r.DataTokens.Any(dt => string.Compare(dt.Key, "area", StringComparison.InvariantCultureIgnoreCase) == 0)))
+                                        || (r.DataTokens != null && r.DataTokens.Any(dt => string.Compare(dt.Key, "area", StringComparison.InvariantCultureIgnoreCase) == 0 
+                                            && string.Compare(dt.Value.ToString(), area, StringComparison.InvariantCultureIgnoreCase) == 0)))
```

Comment:
Sorry, my brain is struggling to make sense of these conditionals. :) What exactly is this doing in plain English? I think this code has reach a level of complexity where it would make sense to refactor some of the conditions into little well-named helper methods that would help readability.